src/game2048/board.rs

```diff
@@ -0,0 +1,317 @@
+use crate::world::Direction;
+use alloc::vec::Vec;
+use rand::prelude::*;
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct World {}
+
+#[derive(Debug)]
+pub struct Tile {
+    pub val: Option<u64>,
+    pub changed: bool,
+    pub row: usize,
+    pub col: usize,
+}
+
+impl Tile {
+    pub fn new(row: usize, col: usize) -> Self {
+        Self {
+            val: None,
+            changed: false,
+            row,
+            col,
+        }
+    }
+
+    fn can_merge(&self, other: &Self) -> bool {
+        if let (Some(ref v1), Some(ref v2)) = (self.val, other.val) {
+            if v1 == v2 {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+pub(crate) struct Board {
+    rows: usize,
+    cols: usize,
+    tiles: Vec<Tile>,
+    rng: rand::rngs::SmallRng,
+}
+
+impl Board {
+    pub fn new(rows: usize, cols: usize) -> Self {
+        let mut tiles = Vec::with_capacity(rows * cols);
+
+        for row in 0..rows {
+            for col in 0..cols {
+                tiles.push(Tile::new(row, col));
+            }
+        }
+
+        Self {
+            rows,
+            cols,
+            tiles,
+            rng: rand::rngs::SmallRng::from_seed([0; 32]),
+        }
+    }
+
+    pub fn reset(&mut self) {
+        for t in self.tiles.iter_mut() {
+            t.val = None;
+            t.changed = true;
+        }
+    }
+
+    pub fn has_changed(&self) -> bool {
+        self.tiles.iter().any(|v| v.changed)
+    }
+
+    pub fn tiles_need_redraw<'a>(&'a self) -> impl Iterator<Item = &'a Tile> {
+        self.tiles.iter().filter(|v| v.changed)
+    }
+
+    pub fn clear_changed(&mut self) {
+        self.tiles.iter_mut().for_each(|v| v.changed = false)
+    }
+
+    pub fn random_fill_empty_tile(&mut self) -> bool {
+        let empty_tiles = self.tiles.iter().filter(|v| v.val.is_none()).count();
+        if empty_tiles == 0 {
+            return false;
+        }
+
+        let random = (self.rng.next_u64() as usize) % empty_tiles;
+        let to_fill = self
+            .tiles
+            .iter_mut()
+            .filter(|v| v.val.is_none())
+            .nth(random)
+            .unwrap();
+        assert!(to_fill.val.is_none());
+        to_fill.val = Some(2);
```

Comment:
I think in the original game there is a 10% chance that this is a 4

Reply:
The english wiki article only mentiones that it is either 2 or 4, but at least the german version of the wiki article speaks about the 10% chance. I'm trying to find a good reference

Reply:
found the ultimate reference: https://github.com/gabrielecirulli/2048/blob/fc1ef4fe5a5fcccea7590f3e4c187c75980b353f/js/game_manager.js#L71

Reply:
fixed